---
title: "Analysis of celiac disease"
author: "Florian Priv√©"
date: "November 3, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "70%", fig.width = 10,
                      fig.asp = 0.7, dev = "svg")
options(width = 95)
```

```{r}
library(tidyverse)
myggplot <- function(..., coeff = 1) {
  bigstatsr:::MY_THEME(ggplot(...), coeff = coeff)
} 
```

## Data

```{r}
# Read file, see "Preprocessing" notebook
library(bigsnpr)
celiac <- snp_attach("backingfiles/celiacQC.rds")
G <- celiac$genotypes
CHR <- celiac$map$chromosome
POS <- celiac$map$physical.pos
NCORES <- nb_cores()
y <- celiac$fam$affection - 1
```

```{r}
# Compute PCA
obj.svd <- snp_autoSVD(G, CHR, POS, ncores = NCORES)
```

## Predictive models

```{r}
# Divide in training/test sets
dim(G)
ind.train <- sort(sample(nrow(G), size = 12e3))
ind.test <- setdiff(rows_along(G), ind.train)
```

### PRS

```{r}
# GWAS
gwas.train <- big_univLogReg(
  G, y[ind.train], ind.train = ind.train, 
  covar.train = obj.svd$u[ind.train, , drop = FALSE], 
  ncores = NCORES
)
snp_qq(gwas.train) +
  coord_cartesian(ylim = c(0, 25))
gwas.train.gc <- snp_gc(gwas.train)
```

```{r, dev="png"}
labels <- 1:22; labels[c(11, 13, 15, 17, 18, 20, 21)] <- ""
cowplot::plot_grid(
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels),
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels) +
    coord_cartesian(ylim = c(0, 25)) + 
    geom_hline(yintercept = -log10(5e-8), color = "red", linetype = 3),
  align = "hv", ncol = 1, labels = LETTERS[1:2], label_size = 25, scale = 0.95
)
```

```{r}
# Clumping on the test set
ind.keep <- snp_clumping(G, infos.chr = CHR,
                         ind.row = ind.test,
                         thr.r2 = 0.2, 
                         S = abs(gwas.train.gc$score),
                         size = 500,
                         is.size.in.bp = TRUE,
                         infos.pos = POS,
                         ncores = NCORES)
```

```{r}
# PRS
thrs <- c(0, -log10(5e-8), exp(seq(log(0.1), log(100), length.out = 100)))
lpS <- -predict(gwas.train.gc)
prs <- snp_PRS(G, betas.keep = gwas.train.gc$estim[ind.keep],
               ind.test = ind.test,
               ind.keep = ind.keep,
               lpS.keep = lpS[ind.keep], 
               thr.list = thrs)
nb.pred <- sapply(thrs, function(thr) sum(lpS[ind.keep] > thr))
aucs <- apply(prs, 2, AUC, target = y[ind.test])
```


### Logistic regression

```{r}
logit <- big_spLogReg(X = G, y01.train = y[ind.train], 
                      ind.train = ind.train, 
                      covar.train = obj.svd$u[ind.train, ],
                      alpha = 0.5, dfmax = 20e3)
preds1 <- predict(logit, X = G, ind.row = ind.test, 
                  covar.row = obj.svd$u[ind.test, ])
library(Matrix)
nb.pred1 <- colSums(logit$beta != 0)
aucs1 <- apply(preds1, 2, AUC, target = y[ind.test])
```

Cross-Model Selection and Averaging (CMSA):

```{r}
system.time({
  cmsa.logit <- big_CMSA(FUN = big_spLogReg, feval = AUC,
                         X = G, y.train = y[ind.train], 
                         ind.train = ind.train, 
                         covar.train = obj.svd$u[ind.train, ],
                         alpha = 0.5, dfmax = 20e3, 
                         ncores = NCORES)
  
  preds2 <- predict(cmsa.logit, X = G, ind.row = ind.test, 
                    covar.row = obj.svd$u[ind.test, ])
})
(auc.cmsa <- AUC(preds2, y[ind.test]))
```


### Logistic regression with feature engineering

```{r}
G2 <- big_attach("backingfiles/celiacQC_tripled1.rds")
logit2 <- big_spLogReg(X = G2, y01.train = y[ind.train], 
                       ind.train = ind.train, 
                       covar.train = obj.svd$u[ind.train, ],
                       alpha = 0.5, dfmax = 20e3)
preds3 <- predict(logit2, X = G2, ind.row = ind.test, 
                  covar.row = obj.svd$u[ind.test, ])
nb.pred2 <- colSums(logit2$beta != 0)
aucs2 <- apply(preds3, 2, AUC, target = y[ind.test])
```

Cross-Model Selection and Averaging (CMSA):

```{r}
system.time({
  cmsa.logit2 <- big_CMSA(FUN = big_spLogReg, feval = AUC,
                          X = G2, y.train = y[ind.train], 
                          ind.train = ind.train, 
                          covar.train = obj.svd$u[ind.train, ],
                          alpha = 0.5, dfmax = 20e3, 
                          ncores = NCORES)
  
  preds4 <- predict(cmsa.logit2, X = G2, ind.row = ind.test, 
                    covar.row = obj.svd$u[ind.test, ])
})
(auc.cmsa2 <- AUC(preds4, y[ind.test]))
```

### Regularization paths

```{r}
h <- round(c(max(aucs), auc.cmsa, auc.cmsa2), 4)
p <- myggplot(rbind(
  cbind.data.frame(nb = nb.pred,      AUC = aucs,      Method = "PRS"),
  cbind.data.frame(nb = nb.pred1[-1], AUC = aucs1[-1], Method = "logit-simple"),
  cbind.data.frame(nb = nb.pred2[-1], AUC = aucs2[-1], Method = "logit-triple")
)) +
  geom_point(aes(nb, AUC, color = Method)) +
  scale_x_log10(breaks = 10^(0:6)) +
  labs(x = "Number of predictors") +
  geom_hline(yintercept = h, color = scales::hue_pal()(3), linetype = 2) +
  theme(legend.position = c(0.85, 0.5))
p + 
  scale_y_continuous(breaks = sort(c(ggplot_build(p)$layout$panel_ranges[[1]]$y.major_source, h)))
```


```{r}
ggsave("figures/main-celiac-regpath.pdf", scale = 1/90, width = 1000, height = 700)
```

### ROC Curves

```{r, fig.asp=1, fig.width=8}
library(plotROC)
scores_tidy <- tibble(
  d = y[ind.test], 
  "PRS-max" =  prs[, which.max(aucs)],
  "logit-simple" = preds2,
  "logit-triple" = preds4
) %>%
  gather(key = "Method", value = "Score", -d)
  
bigstatsr:::MY_THEME(
  ggplot(scores_tidy, aes(d = d, m = Score, color = Method, linetype = Method)) +
    style_roc(xlab = "1 - Specificity", ylab = "Sensitivity")
) +
  geom_roc(n.cuts = 0, size = 2) +
  theme(legend.position = c(0.7, 0.3), legend.key.width = unit(4, "line")) +
  coord_equal()
```

```{r}
ggsave("figures/supp-celiac-roc.pdf", scale = 1/90, width = 700, height = 700)
```

## Top scores

First, we get the SNP with the highest ratio of cases over controls for homozygous-recessive individuals.

```{r}
props <- sapply(cols_along(G), function(j) {
  ccoj <- counts.control[, j]
  mut <- `if`(ccoj[1] < ccoj[3], 1, 3)
  `if`(counts.case[mut, j] > 100, counts.case[mut, j] / ccoj[mut], 0)
})
theone <- which.max(props)
table(Geno = G[, theone], forcats::fct_recode(as.factor(y), Control = "0", Case = "1"))
```

```{r}
# one geno
ind0 <- which(G[ind.test, theone] == 0)
(M <- length(ind0))
mean(y[ind.test[ind0]]) 
# prs max
indPRS <- order(prs[, which.max(aucs)], decreasing = TRUE)[1:M]
mean(y[ind.test[indPRS]])
# logit-simple (CMSA)
indLS <- order(preds2, decreasing = TRUE)[1:M]
mean(y[ind.test[indLS]])
# logit-triple (CMSA)
indLT <- order(preds4, decreasing = TRUE)[1:M]
mean(y[ind.test[indLT]])
```

## Stability of selection of the "logit-simple"

Train on 100 bootstrap samples of the Celiac dataset.

```{r, eval=FALSE}
if (!dir.exists("results6")) dir.create("results6")

for (i in seq_len(100)) {
  
  res.file <- paste0("results6/simu_", i, ".rds")
  if (file.exists(res.file)) next
  
  ind.train <- sort(sample(nrow(G)))
  
  cmsa.logit <- big_CMSA(FUN = big_spLogReg, feval = AUC,
                         X = G, y.train = y[ind.train], 
                         ind.train = ind.train, 
                         covar.train = obj.svd$u[ind.train, ],
                         alpha = 0.5, dfmax = 20e3, 
                         ncores = NCORES)
  
  saveRDS(cmsa.logit[cols_along(G)], file = res.file)
}
```

```{r}
results6 <- list.files("results6", full.names = TRUE) %>%
  map(~readRDS(.x)) %>%
  do.call("cbind", .)
```

```{r}
n_enter <- rowSums(results6 != 0)
mean(n_enter == 0)                ## Proportion of SNPs never entering
n_enter_once <- n_enter[n_enter != 0]
mean(n_enter_once)                ## Mean number of times a SNPs enter 
                                  ## if entering at least once
bigstatsr:::MY_THEME(
  qplot(n_enter_once, geom = "bar") +
    labs(x = "Number of times")
)
```

```{r}
ggsave("figures/supp-celiac-stabilitybar.pdf", scale = 1/90, width = 1000, height = 700)
```

```{r}
length(always_enter <- which(n_enter == ncol(results6)))
cowplot::plot_grid(
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels, 
                ind.highlight = always_enter),
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels, 
                ind.highlight = always_enter) +
    coord_cartesian(ylim = c(0, 25)) + 
    geom_hline(yintercept = -log10(5e-8), color = "red", linetype = 3),
  align = "hv", ncol = 1, labels = LETTERS[1:2], label_size = 25, scale = 0.95
)
```

```{r}
ggsave("figures/supp-celiac-man-col.png", scale = 1/90, width = 1300, height = 1000)
```

```{r}
sq_effects_always <- colSums((results6[always_enter, ])^2)
sq_effects_others <- colSums((results6[-always_enter, ])^2)
summary(100 * sq_effects_always / (sq_effects_always + sq_effects_others))
```

