---
title: "Analysis of celiac disease"
author: "Florian Priv√©"
date: "March 1, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "70%", fig.width = 10,
                      fig.asp = 0.7, dev = "svg", cache = FALSE)
options(width = 95)
```

```{r}
library(tidyverse)
library(bigsnpr)
library(pkg.paper.PRS)
```

## Data

```{r, cache=FALSE}
# Read file, see "Preprocessing" notebook
celiac <- snp_attach("backingfiles/celiacQC.rds")
G <- celiac$genotypes
CHR <- celiac$map$chromosome
POS <- celiac$map$physical.pos
NCORES <- nb_cores()
y <- celiac$fam$affection - 1
obj.svd <- readRDS("backingfiles/PCA.rds")

G2 <- big_attach("backingfiles/celiacQC_tripled1.rds")
```

## Predictive models and plots

```{r, cache=TRUE}
# Divide in training/test sets
dim(G)
set.seed(1)
ind.train <- sort(sample(nrow(G), size = 12e3))
ind.test <- setdiff(rows_along(G), ind.train)
```

### PRS

```{r, cache=TRUE}
# GWAS
system.time(
  gwas.train <- big_univLogReg(
    G, y[ind.train], ind.train = ind.train, 
    covar.train = obj.svd$u[ind.train, , drop = FALSE], 
    ncores = NCORES
  )
)
```

```{r, dev='png', fig.asp=0.8, cache=TRUE}
# Q-Q plot
snp_qq(gwas.train) +
  coord_cartesian(ylim = c(0, 15))
gwas.train.gc <- snp_gc(gwas.train)
```

```{r, dev="png", fig.asp=0.9, out.width="90%", cache=TRUE}
# Manhattan plot
labels <- 1:22; labels[c(11, 13, 15, 17, 18, 20, 21)] <- ""
cowplot::plot_grid(
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels),
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels) +
    coord_cartesian(ylim = c(0, 22)) + 
    geom_hline(yintercept = -log10(5e-8), color = "red", linetype = 3),
  align = "hv", ncol = 1, labels = LETTERS[1:2], label_size = 25, scale = 0.95
)
```

```{r, cache=TRUE}
# Clumping on the test set
ind.keep <- snp_clumping(G, infos.chr = CHR,
                         ind.row = ind.test,
                         thr.r2 = 0.2, 
                         S = abs(gwas.train.gc$score),
                         size = 500,
                         is.size.in.bp = TRUE,
                         infos.pos = POS,
                         ncores = NCORES)
```

```{r, cache=TRUE}
# PRS
thrs <- c(0, -log10(5e-8), exp(seq(log(0.1), log(100), length.out = 100)))
lpS <- -predict(gwas.train.gc)
prs <- snp_PRS(G, betas.keep = gwas.train.gc$estim[ind.keep],
               ind.test = ind.test,
               ind.keep = ind.keep,
               lpS.keep = lpS[ind.keep], 
               thr.list = thrs)
nb.pred <- sapply(thrs, function(thr) sum(lpS[ind.keep] > thr))
aucs <- apply(prs, 2, AUC, target = y[ind.test])
summary(aucs)
nb.pred[which.max(aucs)]
```

```{r}
pAUC <- pkg.paper.PRS:::pAUC

plot_density_AUC <- function(pred, true = y[ind.test]) {
  true_fct <- forcats::fct_recode(as.factor(true), "Control" = "0", "Case" = "1")
  data.frame(Score = pred, Status = true_fct) %>%
  ggplot() +
  geom_density(aes(Score, fill = Status), alpha = 0.4) +
  ggtitle(sprintf("AUC = %s%% | pAUC = %s", 
                  round(100 * AUC(pred, true), 2),
                  round(pAUC(pred, true), 4))) + 
    theme_bigstatsr() +
    theme(plot.title = element_text(size = rel(1.5), hjust = 0.5))
}
(p1 <- plot_density_AUC(prs[, which.max(aucs)]))
```


### Logistic regression

Cross-Model Selection and Averaging (CMSA):

```{r, cache=TRUE}
# Train logit-simple
system.time(
  logit <- big_spLogReg(X = G, y01.train = y[ind.train], 
                        ind.train = ind.train, 
                        covar.train = obj.svd$u[ind.train, ],
                        ncores = NCORES, alphas = 1)
)
# Get K = 10 predictions/individuals
preds1 <- predict(logit, X = G, ind.row = ind.test, 
                  covar.row = obj.svd$u[ind.test, ])
(nb.pred1 <- sum(rowSums(sapply(logit, function(l) l$beta.X != 0)) > 0))
(aucs1 <- AUC(preds1, target = y[ind.test]))
```

```{r}
(p2 <- plot_density_AUC(preds1))
```

### Logistic regression with feature engineering

```{r, cache=TRUE}
system.time(
  logit2 <- big_spLogReg(X = G2, y01.train = y[ind.train], 
                         ind.train = ind.train, 
                         covar.train = obj.svd$u[ind.train, ],
                         ncores = NCORES, alphas = 1)
)

preds2 <- predict(logit2, X = G2, ind.row = ind.test, 
                  covar.row = obj.svd$u[ind.test, ])
(nb.pred2 <- sum(rowSums(sapply(logit2, function(l) l$beta.X != 0)) > 0))
(aucs2 <- AUC(preds2, target = y[ind.test]))
```

```{r}
(p3 <- plot_density_AUC(preds2))
```

```{r, out.width="70%", fig.asp=900/780}
list(p1, p2, p3) %>%
  lapply(function(p) p + theme(legend.position = "none")) %>%
  cowplot::plot_grid(plotlist = ., ncol = 1, labels = LETTERS[1:3], label_size = 20, scale = 0.95) %>%
  cowplot::plot_grid(., cowplot::get_legend(p1), rel_widths = c(1, 0.15), scale = 0.95)
ggsave("figures/supp-score-densities.pdf", scale = 1/100, width = 780, height = 900)
```


### ROC Curves

```{r, fig.asp=1, fig.width=8}
library(plotROC)
scores_tidy <- tibble(
  d = y[ind.test], 
  "PRS-max"      = prs[, which.max(aucs)],
  "logit-simple" = preds1,
  "logit-triple" = preds2
) %>%
  gather(key = "Method", value = "Score", -d)
  
methods.color <- setNames(scales::hue_pal()(6), 
                          c("logit-simple", "PRS-max", "PRS-stringent",
                            "T-Trees", "logit-triple", "PRS-all")) 

ggplot(scores_tidy, aes(d = d, m = Score, color = Method, linetype = Method)) +
  style_roc(xlab = "1 - Specificity", ylab = "Sensitivity") +
  theme_bigstatsr() +
  geom_roc(n.cuts = 0, size = 2) +
  theme(legend.position = c(0.7, 0.3), legend.key.width = unit(4, "line")) +
  coord_equal() +
  scale_color_manual(values = methods.color)
```

```{r}
ggsave("figures/celiac-roc.pdf", scale = 1/90, width = 700, height = 700)
```


## Predictive methods (multiple times)

### Methods' functions

Each method's function returns a tibble (data frame) with 4 columns:

1. the name of the method,
2. the predictive scores and true phenotypes for the test set, as a [list-column](http://r4ds.had.co.nz/many-models.html),
3. the timing of the main computations (in seconds),
4. the number of SNPs used for the prediction.

### Run multiple times

```{r, eval=TRUE}
if (!dir.exists("results6")) dir.create("results6")
pheno.all <- y
covar.all <- obj.svd$u

for (i in 1:100) {
  
  res.file <- paste0("results6/simu_", i, ".rds")
  if (file.exists(res.file)) next
  
  # Split in training/test sets
  ind.train <- sort(sample(nrow(G), size = 12e3))
  ind.test <- setdiff(rows_along(G), ind.train)
    
  # Get results from all methods
  res <- bind_rows(
    filter(PRS(G, CHR, POS, pheno.all, covar.all, ind.train, ind.test), thr.r2 == 0.2),
    logit.CMSA(G,  pheno.all, covar.all, ind.train, ind.test, "logit-simple"),
    logit.CMSA(G2, pheno.all, covar.all, ind.train, ind.test, "logit-triple")
  )
  res$pheno <- list(pheno.all[ind.test])
  
  saveRDS(res, file = res.file)
}
```

```{r}
list.files(path = "results6/", full.names = TRUE) %>%
  map_df(readRDS) %>%
  as_tibble() %>%
  mutate(
    AUC = map2_dbl(pred, pheno, ~bigstatsr::AUC(.x, .y)),
    pAUC = map2_dbl(pred, pheno, ~pAUC(.x, .y)),
    nb.preds = map_int(set, length)
  ) %>%
  group_by(method) %>%
  summarize_at(c("AUC", "pAUC", "nb.preds", "timing"), function(x) {
    glue::glue("{signif(mean(x), 3)} ({signif(boot(x), 3)})")
  }) %>%
  print() %>%
  xtable::xtable()
```



