---
title: "Analysis of celiac disease"
author: "Florian Priv√©"
date: "November 3, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "70%", 
                      fig.asp = 0.7, error = TRUE)
options(width = 85)
```

```{r}
library(ggplot2)
myggplot <- function(..., coeff = 1) {
  bigstatsr:::MY_THEME(ggplot(...), coeff = coeff)
} 
```

## Data

```{r}
# Read file, see "Preprocessing" notebook
celiac <- snp_attach("backingfiles/celiacQC.rds")
G <- celiac$genotypes
CHR <- celiac$map$chromosome
POS <- celiac$map$physical.pos
NCORES <- nb_cores()
y <- celiac$fam$affection - 1
```

```{r}
# Compute PCA
obj.svd <- snp_autoSVD(G, CHR, POS, ncores = NCORES)
```

## Predictive models

```{r}
# Divide in training/test sets
dim(G)
ind.train <- sort(sample(nrow(G), size = 12e3))
ind.test <- setdiff(rows_along(G), ind.train)
```

### PRS

```{r}
# GWAS
gwas.train <- big_univLogReg(
  G, y[ind.train], ind.train = ind.train, 
  covar.train = obj.svd$u[ind.train, , drop = FALSE], 
  ncores = NCORES
)
gwas.train.gc <- snp_gc(gwas.train)
```

```{r}
labels <- 1:22; labels[c(11, 13, 15, 17, 18, 20, 21)] <- ""
cowplot::plot_grid(
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels),
  snp_manhattan(gwas.train.gc, CHR, POS, labels = labels) +
    coord_cartesian(ylim = c(0, 30)) + 
    geom_hline(yintercept = -log10(5e-8), color = "red", linetype = 3),
  align = "hv", ncol = 1, labels = LETTERS[1:2], label_size = 25, scale = 0.95
)
```

```{r}
ggsave("figures/celiac-man.png", scale = 1/90, width = 1400, height = 1000)
```

```{r}
# Clumping on the test set
ind.keep <- snp_clumping(G, infos.chr = CHR,
                         ind.row = ind.test,
                         thr.r2 = 0.2, 
                         S = abs(gwas.train.gc$score),
                         size = 500,
                         is.size.in.bp = TRUE,
                         infos.pos = POS,
                         ncores = NCORES)
```

```{r}
# PRS
thrs <- c(0, -log10(5e-8), exp(seq(log(0.1), log(100), length.out = 100)))
lpS <- -predict(gwas.train.gc)
prs <- snp_PRS(G, betas.keep = gwas.train.gc$estim[ind.keep],
               ind.test = ind.test,
               ind.keep = ind.keep,
               lpS.keep = lpS[ind.keep], 
               thr.list = thrs)
nb.pred <- sapply(thrs, function(thr) sum(lpS[ind.keep] > thr))
aucs <- apply(prs, 2, AUC, target = y[ind.test])
```


### Logistic regression

```{r}
logit <- big_spLogReg(X = G, y01.train = y[ind.train], 
                      ind.train = ind.train, 
                      covar.train = obj.svd$u[ind.train, ],
                      alpha = 0.5, dfmax = 20e3)
preds1 <- predict(logit, X = G, ind.row = ind.test, 
                  covar.row = obj.svd$u[ind.test, ])
library(Matrix)
nb.pred1 <- colSums(logit$beta != 0)
aucs1 <- apply(preds1, 2, AUC, target = y[ind.test])
```

Cross-Model Selection and Averaging:

```{r}
system.time({
  cmsa.logit <- big_CMSA(FUN = big_spLogReg, feval = AUC,
                         X = G, y.train = y[ind.train], 
                         ind.train = ind.train, 
                         covar.train = obj.svd$u[ind.train, ],
                         alpha = 0.5, dfmax = 20e3, 
                         ncores = NCORES)
  
  preds2 <- predict(cmsa.logit, X = G, ind.row = ind.test, 
                    covar.row = obj.svd$u[ind.test, ])
})
(auc.cmsa <- AUC(preds2, y[ind.test]))
```


### Logistic regression with feature engineering

```{r}
G2 <- big_attach("backingfiles/celiacQC_tripled1.rds")
logit2 <- big_spLogReg(X = G2, y01.train = y[ind.train], 
                       ind.train = ind.train, 
                       covar.train = obj.svd$u[ind.train, ],
                       alpha = 0.5, dfmax = 20e3)
preds3 <- predict(logit2, X = G2, ind.row = ind.test, 
                  covar.row = obj.svd$u[ind.test, ])
nb.pred2 <- colSums(logit2$beta != 0)
aucs2 <- apply(preds3, 2, AUC, target = y[ind.test])
```

Cross-Model Selection and Averaging:

```{r}
system.time({
  cmsa.logit2 <- big_CMSA(FUN = big_spLogReg, feval = AUC,
                          X = G2, y.train = y[ind.train], 
                          ind.train = ind.train, 
                          covar.train = obj.svd$u[ind.train, ],
                          alpha = 0.5, dfmax = 20e3, 
                          ncores = NCORES)
  
  preds4 <- predict(cmsa.logit2, X = G2, ind.row = ind.test, 
                    covar.row = obj.svd$u[ind.test, ])
})
(auc.cmsa2 <- AUC(preds4, y[ind.test]))
```

### Regularization paths

```{r}
h <- round(c(max(aucs), auc.cmsa, auc.cmsa2), 4)
p <- myggplot(rbind(
  cbind.data.frame(nb = nb.pred,      AUC = aucs,      Method = "PRS"),
  cbind.data.frame(nb = nb.pred1[-1], AUC = aucs1[-1], Method = "logit-simple"),
  cbind.data.frame(nb = nb.pred2[-1], AUC = aucs2[-1], Method = "logit-triple")
)) +
  geom_point(aes(nb, AUC, color = Method)) +
  scale_x_log10(breaks = 10^(0:6)) +
  labs(x = "Number of predictors") +
  geom_hline(yintercept = h, color = scales::hue_pal()(3), linetype = 2) +
  theme(legend.position = c(0.85, 0.5))
p + 
  scale_y_continuous(breaks = sort(c(ggplot_build(p)$layout$panel_ranges[[1]]$y.major_source, h)))
```


```{r}
ggsave("figures/celiac-regpath.pdf", scale = 1/90, width = 1000, height = 700)
```
