---
title: "Simulations"
author: "Florian Privé"
date: "October 20, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.align = "center", out.width = "70%", 
                      fig.asp = 0.7)
options(width = 85)
```

## Methods' functions

```{r, message=FALSE}
library(tidyverse)
library(glue)
```

Each method's function returns a tibble (data frame) with 4 columns:

1. the name of the method,
2. the predictive scores and true phenotypes for the test set, as a [list-column](http://r4ds.had.co.nz/many-models.html),
3. the timing of the main computations (in seconds),
4. the number of SNPs used for the prediction.

### Logistic regression

```{r}
logit.CMSA <- function(G, pheno.all, covar.all, ind.train, ind.test, method) {
  
  timing <- system.time({
    
    cmsa.logit <- big_spLogReg(X = G, y01.train = pheno.all[ind.train], 
                               ind.train = ind.train, 
                               covar.train = covar.all[ind.train, , drop = FALSE],
                               ncores = nb_cores())
    
    preds <- rowMeans(
      predict(cmsa.logit, X = G, ind.row = ind.test, 
                     covar.row = covar.all[ind.test, , drop = FALSE])
    )
  })[3]
  
  tibble(
    method   = method, 
    eval     = list(cbind(preds, pheno.all[ind.test])),
    timing   = timing,
    nb.preds = sum(rowSums(sapply(cmsa.logit, function(x) x$beta.X) != 0) > 0)
  )
}
```

## Simulations

### Data

```{r}
library(bigsnpr)
celiac2 <- snp_attach("backingfiles/celiacQC.rds")
CHR <- celiac2$map$chromosome
POS <- celiac2$map$physical.pos
G <- celiac2$genotypes

covar.all <- readRDS("backingfiles/PCA.rds")$u

n <- nrow(G)
```

### Simulation of phenotypes

```{r}
get_pheno <- function(
  G,                                        ## matrix of genotypes
  h2,                                       ## heritability 
  M,                                        ## nbs of causal variants
  ind.possible = cols_along(G),             ## indices of possible causal variants
  effects.dist = c("gaussian", "laplace"),  ## distribution of effects 
  model = c("simple", "fancy"),             ## model for simulation
  K = 0.3                                   ## prevalence
) {
  
  effects.dist  <- match.arg(effects.dist)
  model <- match.arg(model)
  
  set <- sample(ind.possible, size = M)
  effects <- `if`(effects.dist == "gaussian", 
                  rnorm(M, sd = sqrt(h2 / M)),
                  rmutil::rlaplace(M, s = sqrt(h2 / (2*M))))
  
  if (model == "simple") {
    # only linear
    y.simu <- scale(G[, set]) %*% effects
  } else {
    sets <- split(1:M, sample(rep_len(1:3, M)))
    # linear
    ind1 <- sets[[1]]
    y.simu <- scale(G[, set[ind1]]) %*% effects[ind1]
    # recessive / dominant
    ind2 <- sets[[2]]
    y.simu <- y.simu + scale(G[, set[ind2]] > 0.5) %*% effects[ind2]
    # interactions
    ind3 <- matrix(sets[[3]], ncol = 2)
    y.simu <- y.simu + scale(G[, set[ind3[, 1]]] * G[, set[ind3[, 2]]]) %*% 
      effects[ind3[, 1]] * sqrt(2)
  }
  
  y.simu <- y.simu / sd(y.simu) * sqrt(h2)
  stopifnot(all.equal(drop(var(y.simu)), h2))
  y.simu <- y.simu + rnorm(nrow(G), sd = sqrt(1 - h2))
  pheno <- as.numeric(y.simu > qnorm(1 - K))
}
```


### Scenario n°1 (without T-Trees)

```{r}
params.grid2 <- expand.grid(
  n.train = 6000,
  par.causal = list(c(30, "all"), c(300, "all"), c(3000, "all"), c(30, "HLA")), 
  par.dist   = c("gaussian", "laplace"), 
  par.h2     = c(0.5, 0.8), 
  par.model  = c("simple", "fancy"),
  num.simu   = 1:100,
  stringsAsFactors = FALSE
) 
```

```{r, eval=TRUE}
if (!dir.exists("results2")) dir.create("results2")

for (i in rows_along(params.grid2)) {
  
  res.file <- paste0("results2/simu_", i, ".rds")
  if (file.exists(res.file)) next
  
  params <- params.grid2[i, ]
  par.causal <- params[["par.causal"]][[1]]
  
  # Simulate phenotypes
  pheno.all <- get_pheno(
    G,    
    h2 = params[["par.h2"]], 
    M = as.integer(par.causal[1]), 
    ind.possible = `if`(par.causal[2] == "all", cols_along(G), ind.HLA),
    effects.dist = params[["par.dist"]], 
    model = params[["par.model"]], 
    K = 0.3                 
  )
  
  # Split in training/test sets
  ind.train <- sort(sample(n, size = params[["n.train"]]))
  ind.test <- setdiff(1:n, ind.train)
    
  # Get results from all methods
  res <- bind_rows(
    PRS(G, CHR, POS, pheno.all, covar.all, ind.train, ind.test),
    logit.CMSA(G,  pheno.all, covar.all, ind.train, ind.test, "logit-simple"),
    logit.CMSA(G2, pheno.all, covar.all, ind.train, ind.test, "logit-triple")
  )
  params[["res"]] <- list(res)
  saveRDS(unnest(params, res, .drop = FALSE), file = res.file)
}
```


### Scenario n°2 (dataset with only chromosome 6)

```{r}
params.grid4 <- expand.grid(
  n.train    = 6000,
  par.causal = list(c(30, "all"), c(300, "all"), c(3000, "all"), c(30, "HLA")), 
  par.dist   = c("gaussian", "laplace"), 
  par.h2     = c(0.5, 0.8), 
  par.model  = "simple",
  num.simu   = 1:100,
  stringsAsFactors = FALSE
) 
```

```{r, eval=TRUE}
if (!dir.exists("results4")) dir.create("results4")

G6 <- big_copy(G, ind.col = which(CHR == 6))
ind.HLA6 <- snp_indLRLDR(CHR[CHR == 6], POS[CHR == 6], 
                         subset(LD.wiki34, ID == "hild12"))

for (i in rows_along(params.grid4)) {
  
  res.file <- paste0("results4/simu_", i, ".rds")
  if (file.exists(res.file)) next
  
  params <- params.grid4[i, ]
  par.causal <- params[["par.causal"]][[1]]
  
  # Simulate phenotypes
  pheno.all <- get_pheno(
    G6,    
    h2 = params[["par.h2"]], 
    M = as.integer(par.causal[1]), 
    ind.possible = `if`(par.causal[2] == "all", cols_along(G6), ind.HLA6),
    effects.dist = params[["par.dist"]], 
    model = params[["par.model"]], 
    K = 0.3                 
  )
  
  # Split in training/test sets
  ind.train <- sort(sample(n, size = params[["n.train"]]))
  ind.test <- setdiff(1:n, ind.train)
    
  # Get results from all methods
  res <- bind_rows(
    PRS(G6, CHR[CHR == 6], POS[CHR == 6], pheno.all, covar.all, ind.train, ind.test),
    logit.CMSA(G6, pheno.all, covar.all, ind.train, ind.test, "logit-simple")
  )
  params[["res"]] <- list(res)
  saveRDS(unnest(params, res, .drop = FALSE), file = res.file)
}
```

### Scenario n°3 (varying training size)

```{r}
params.grid5 <- expand.grid(
  n.train    = 1:5 * 1000,
  par.causal = list(c(300, "all")), 
  par.dist   = c("gaussian", "laplace"), 
  par.h2     = c(0.5, 0.8), 
  par.model  = "simple",
  num.simu   = 1:100,
  stringsAsFactors = FALSE
) 
```

```{r, eval=TRUE}
if (!dir.exists("results5")) dir.create("results5")

for (i in rows_along(params.grid5)) {
  
  res.file <- paste0("results5/simu_", i, ".rds")
  if (file.exists(res.file)) next
  
  params <- params.grid5[i, ]
  par.causal <- params[["par.causal"]][[1]]
  
  # Simulate phenotypes
  pheno.all <- get_pheno(
    G,    
    h2 = params[["par.h2"]], 
    M = as.integer(par.causal[1]), 
    ind.possible = `if`(par.causal[2] == "all", cols_along(G), ind.HLA),
    effects.dist = params[["par.dist"]], 
    model = params[["par.model"]], 
    K = 0.3                 
  )
  
  # Split in training/test sets
  ind.train <- sort(sample(n, size = params[["n.train"]]))
  ind.test <- setdiff(1:n, ind.train)
    
  # Get results from all methods
  res <- bind_rows(
    PRS(G, CHR, POS, pheno.all, covar.all, ind.train, ind.test),
    logit.CMSA(G,  pheno.all, covar.all, ind.train, ind.test, "logit-simple")
  )
  params[["res"]] <- list(res)
  saveRDS(unnest(params, res, .drop = FALSE), file = res.file)
}
```